Create Dockerfiles for Slim and Alpine builds

jeancochrane opened the pull request:
## Overview

Define two Dockerfile templates for Slim and Alpine builds, each of which compiles GDAL from source using Java and Python bindings. Bootstrap the Dockerfile templates off of https://github.com/geotrellis/geotrellis-gdal, with minor modifications to make builds faster.

Closes https://github.com/geotrellis/geotrellis-server/issues/60.

## Notes

* We had debated switching to an approach that would use the prebuilt [`libgdal-java`](https://packages.debian.org/stretch/libgdal-java) package to install these bindings. While this approach would benefit from faster build times and less maintenance on our end, I moved away from it for a few reasons:
    * `libgdal-java` is not available on `apk`
    * Nathan and Grigory both mentioned wanting to use the latest version of Geotrellis, which is only available on `experimental` Debian builds
    * After an hour or so of tweaking I couldn't get the package installation to pass the tests in https://github.com/geotrellis/geotrellis-server using a `sid-slim` build

## Testing instructions

* Confirm that Slim and Alpine builds compile successfully on Travis
* Run `GDAL_VERSION=2.3.2 OPENJPEG_VERSION=2.3.0 VARIANT=slim ./scripts/cibuild` to build a Slim image (note that this can take ~30 minutes, so plan ahead)
* Clone https://github.com/geotrellis/geotrellis-server and `cd` into it
* Run the following commands to test the images against real Geotrellis tests:

```
docker run \
  -v $(pwd)/build.sbt:/root/build.sbt \
  -v $(pwd)/project:/root/project \
  -v $(pwd)/core:/root/core \
  -v $(pwd)/sbt:/root/sbt \
  -w /root \
  quay.io/azavea/openjdk-gdal:2.3.2-slim \
  ./sbt "project core" test
```

* Clone https://github.com/geotrellis/geotrellis-gdal/ and `cd` into it
* Run the following commands to test the images against more real Geotrellis tests:

```
docker run \
  -v $(pwd)/build.sbt:/root/build.sbt \
  -v $(pwd)/project:/root/project \
  -v $(pwd)/gdal:/root/gdal \
  -v $(pwd)/src:/root/src \
  -v $(pwd)/sbt:/root/sbt \
  -w /root \
  quay.io/azavea/openjdk-gdal:2.3.2-slim \
  ./sbt "project gdal" test
```

moradology committed:
Test in dockerfile to avoid dep pain

(cherry picked from commit c9e07a1616b11041528c6847db1106cc89d5839e)

jbouffard committed:
Fixed how Travis runs the unit tests

Fixed typo in Travis script

Changed how Travis runs the tests

(cherry picked from commit 01ccb972dbe194d0ec0299cd8c75d50f0bcb592b)

Commit:
Update versions to most recent

(cherry picked from commit fbcab0e4ad1ce6f24a7806e742a122c5191e0344)

pomadchin committed:
Update deps, and run GDAL tests in travis

(cherry picked from commit 61a4fe777cedddab06312041ec73f6345874ab16)

pomadchin committed:
Fix Dockerfile to install JNI bindings

(cherry picked from commit a31e3151d86231fed0482fb0bbc91685dc01a45a)

jbouffard committed:
Changed what *.so files are copied over when installing the JNI bindings in the Docker container

(cherry picked from commit da5dacf0f96fab78187733d7545bde77ad707ae9)

pomadchin committed:
Update Dockerfile, add Dockerfile.alpine, bump up gdal, java, set ld_path

(cherry picked from commit 38b62bf545d734cd738820431f21d21373355ea8)

Commit:
Run Make in parallel and clean up dependencies

Use three jobs for all make builds in order to speed up the compile
time. In addition, make sure to compile GDAL with poppler support,
and remove an extraneous libspatialite-dev dependency from the list of
apt dependencies.

Commit:
Rename Dockerfile.alpine

The alpine image had a typo in the name. Rename the file to use
the correct spelling.

Commit:
Add LICENSE and README

Add generic placeholder files for an Apache2 license and a README.

Commit:
Parallelize Make compilation in Alpine image

As in 255640113a6219e9315fee9204f649d3436b6108, run Make in parralel
for Alpine builds.

Commit:
Reorganize repo for CI

Move Alpine and Slim Dockerfiles into templates with variables for
GDAL_VERSION and OPENJPEG_VERSION. In addition, scaffold a .travis.yml
file to make use of cibuild and test scripts.

Commit:
Add deploy step to Travis config

Add a deploy step directing Travis to publish container images
to Quay.

Commit:
Remove stray Dockerfile and add docs

Flesh out the README with instructions on building and testing images.
In addition, remove a stray Dockerfile that got created by 'cibuild'
and add a .gitignore file that ignores Dockerfiles.

Commit:
Fix versioning errors in Dockerfile-alpine.template

Previously, the Alpine build template had hardcoded the version
strings for GDAL, OpenJPEG, and Ant. Update the file to accept
version strings as variables during 'cibuild' instead.

jeancochrane commented on Dockerfile-alpine.template:
```diff
@@ -0,0 +1,69 @@
+ENV ROOTDIR /usr/local/
+ENV LD_LIBRARY_PATH /usr/local/lib
+ENV GDAL_VERSION %%GDAL_VERSION%%
+ENV OPENJPEG_VERSION %%OPENJPEG_VERSION%%
```
I'm not totally sure why we're compiling OpenJPEG from source, although I assume there's a version mismatch with GDAL. @moradology may have more context here.

jeancochrane commented on Dockerfile-alpine.template:
```diff
@@ -0,0 +1,69 @@
+
+ENV ANT_VERSION %%ANT_VERSION%%
+ENV ANT_HOME $ROOTDIR/src/apache-ant-${ANT_VERSION}/
+ENV PATH ${ANT_HOME}/bin:${PATH}
```
Ant is necessary to compile Java bindings, and is not available on `apk`.

hectcastro replied:
Could we use [`apache-ant`](https://pkgs.alpinelinux.org/package/edge/community/x86/apache-ant)?

jeancochrane commented on Dockerfile-alpine.template:
```diff
@@ -0,0 +1,69 @@
+RUN cd src && tar -xvf openjpeg-${OPENJPEG_VERSION}.tar.gz && cd openjpeg-${OPENJPEG_VERSION}/ \
+    && mkdir build && cd build \
+    && cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=$ROOTDIR \
+    && make -j3 && make -j3 install && make -j3 clean \
```
`make -j3` gives us up to three parallel jobs in Make. Xenial builds on Travis only have two cores available, but the docs mention [you can add extra](https://docs.travis-ci.com/user/speeding-up-the-build/#makefile-optimization) for heavy I/O (not sure how this actually works).

Note that the [GDAL CI pipeline uses `-j3` on Travis](https://github.com/OSGeo/gdal/blob/463e28a58d73e5f7bc4f554d999b2bb2e9bab799/gdal/ci/travis/ubuntu_1604/install.sh#L12).

jeancochrane commented on Dockerfile-alpine.template:
```diff
@@ -0,0 +1,69 @@
+    && cd $ROOTDIR && cd src/gdal-${GDAL_VERSION}/swig/java && cp -f gdal.jar /usr/local/lib/ \
+    && cd $ROOTDIR && rm -Rf src/gdal*
+
+RUN touch ${JAVA_HOME}/release
```
Not really sure why we need to `touch` this release file.

hectcastro replied:
If @pomadchin doesn't know, then we should consider removing it.

hectcastro commented on LICENSE:
```diff
@@ -0,0 +1,202 @@
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright 2017 Azavea Inc.
```
Reminder to update this to 2019 before merging.

hectcastro commented on scripts/cipublish:
```diff
@@ -0,0 +1,21 @@
+    else
+        docker login -u "${QUAY_USER}" -p "${QUAY_PASSWORD}" quay.io
+
+        docker push "quay.io/azavea/openjdk-gdal:$GDAL_VERSION-$VARIANT"
```
I created a stub repository in Quay for this at https://quay.io/repository/azavea/openjdk-gdal. We'll still need to get the credentials for the `azavea+images` robot and feed them to Travis CI so that the login command above can succeed.

jeancochrane replied:
I temporarily set Travis to deploy from this branch in https://github.com/azavea/docker-openjdk-gdal/pull/1/commits/85a30ec017038408e2d4ec89a57d272a95b93006 to make sure that this would work. Images are up here: https://quay.io/repository/azavea/openjdk-gdal The commit has been reverted, so we should only deploy from `master` now.

hectcastro commented on README.md:
````diff
@@ -0,0 +1,54 @@
+```
+
+This image exists primarily to help bundle GDAL for use with
+[Geotrellis](https://github.com/geotrellis/geotrellis-server). To run Geotrellis
````
s/Geotrellis/GeoTrellis/

hectcastro commented on Dockerfile-slim.template:
```diff
@@ -0,0 +1,66 @@
+FROM openjdk:8u181
```
I think that we should try to make this more generic, like the Alpine image (e.g., `openjdk:8`).

As a side note, I tested this with a JRE based image vs. a JDK based one, and it fails to build (JRE image is usually a lot smaller, which is why we aim to use it when we can).

jeancochrane replied:
Do you think it's worth trying to get the JRE image to build? Seems worth a try, but it'll likely take some time to figure out.

hectcastro replied:
Probably not. It may be doable, but there seem to be enough other fish to fry here.

hectcastro commented on Dockerfile-alpine.template:
```diff
@@ -0,0 +1,69 @@
+# Load assets
+WORKDIR $ROOTDIR/
+
+ADD http://download.osgeo.org/gdal/${GDAL_VERSION}/gdal-${GDAL_VERSION}.tar.gz $ROOTDIR/src/
```
> Because image size matters, using `ADD` to fetch packages from remote URLs is strongly discouraged; you should use `curl` or `wget` instead. That way you can delete the files you no longer need after they’ve been extracted and you don’t have to add another layer in your image.

See: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy

I would consider doing this with something like:

```
wget -qO- http://download.osgeo.org/gdal/${GDAL_VERSION}/gdal-${GDAL_VERSION}.tar.gz \
| tar -xzC $ROOTDIR/src/ \
&& cd $ROOTDIR/src/
```

jeancochrane replied:
That's awesome context 👍

hectcastro commented on Dockerfile-alpine.template:
```diff
@@ -0,0 +1,69 @@
+ADD https://www.apache.org/dist/ant/binaries/apache-ant-${ANT_VERSION}-bin.tar.gz $ROOTDIR/src/ant-${ANT_VERSION}.tar.gz
+
+# Install basic dependencies
+RUN apk update && apk add \
```
I think that this use of `apk` should be modified to remove build dependencies after we're done with them. Unfortunately, that might require sifting through these to determine which are true build dependencies vs. runtime ones.

https://github.com/azavea/docker-flask/blob/f9433b9a4d62336d9914b386ef9b178f93630c92/Dockerfile-alpine.template#L7-L12

(Note how the `pip install` step that needs to build dependencies is wrapped by the install and uninstall steps.)

hectcastro commented on Dockerfile-alpine.template:
```diff
@@ -0,0 +1,69 @@
+# Deps required for GDAL JNI bindings
+RUN apk add swig
+
+RUN apk update
```
It is unclear what this stage is for.

hectcastro commented on Dockerfile-alpine.template:
```diff
@@ -0,0 +1,69 @@
+RUN touch ${JAVA_HOME}/release
+
+# Output version and capabilities by default.
+CMD gdalinfo --version && gdalinfo --formats && ogrinfo --formats
```
Not sure this is something we want as the default `CMD`, especially for instances where we are using this to run a GeoTrellis service. I think it would be better to leave this stage out and bake it into `test`.

hectcastro commented on Dockerfile-slim.template:
```diff
@@ -0,0 +1,66 @@
+ADD https://github.com/uclouvain/openjpeg/archive/v${OPENJPEG_VERSION}.tar.gz $ROOTDIR/src/openjpeg-${OPENJPEG_VERSION}.tar.gz
+
+# Install basic dependencies
+RUN apt-get update -y && apt-get install -y \
```
Similar comment here as above regarding build dependencies:

https://github.com/azavea/docker-flask/blob/f9433b9a4d62336d9914b386ef9b178f93630c92/Dockerfile-slim.template#L7-L14

hectcastro reviewed (commented):
Left most of my comments on the Alpine `Dockerfile`, but a bunch apply to the Debian one too.

Commit:
Split out build and runtime dependencies

Separate build and runtime dependencies during package installation
and compilation so that build dependencies can be deleted from the
image. Also, do a little bit of related Dockerfile tidying.

Commit:
Temporarily deploy to Quay from feature branch

Test that Travis can deploy images to Quay.io by temporarily setting
the 'branch' to 'feature/jfc/port-dockerfiles'.

This commit will be reverted before merging.

Commit:
Make scripts/cipublish executable

Previously, 'scripts/cipublish' did not have Unix permissions
for execution. Make it executable so Travis can run it.

Commit:
Revert "Temporarily deploy to Quay from feature branch"

Deploy new container images from the 'master' branch.

This reverts commit 85a30ec017038408e2d4ec89a57d272a95b93006.

jeancochrane commented:
Alright, this should be ready for another pass.

hectcastro commented on scripts/cibuild:
```diff
@@ -0,0 +1,32 @@
+            -e "s/%%ANT_VERSION%%/$ANT_VERSION/" \
+            "Dockerfile-$VARIANT.template" > Dockerfile
+
+        docker build -t "quay.io/azavea/openjdk-gdal:$GDAL_VERSION-$VARIANT" .
```
I think that we need to make the tag a bit more descriptive, which could mean that we need to parametrize OpenJDK version, but I could also see just leaving it as is, but also hardcoding the version into the tag name like:

```
quay.io/azavea/openjdk-gdal:8-jdk-gdal-2.3
```

Not pretty, but I think we need a way to surface both the underlying JDK and GDAL versions.

jeancochrane replied:
How about:

```
quay.io/azavea/openjdk-gdal:2.3.2-jdk8-slim
```

So that we can leverage the implied GDAL versioning in `gdal:2.3.2`? That seems to be similar to the pattern we follow with our Django images: https://quay.io/repository/azavea/django?tag=latest&tab=tags

hectcastro replied:
Heh, I went back-and-forth on it and settled on putting the JDK version first because it is first in the container image name. I also used `8-jdk` to align with the official container image naming structure.

This might be a thing where you propose both options to 3 people, see what makes sense to the majority, and just go with that.

hectcastro commented on Dockerfile-alpine.template:
```diff
@@ -0,0 +1,59 @@
+    tar -xzC $ROOTDIR/src/
+
+RUN set -ex \
+    && apk update \
```
I think that you can use the `--update` flag for `apk add` vs. having to do a standalone `update` here.

hectcastro reviewed (commented):
Changes here all look good, but now my `geotrellis-gdal` tests are failing. One of the output messages is:

```
[info] - Java GDAL bindings not installed, skipping !!! IGNORED !!!
```

Failures are happening for both the `slim` and `alpine` variants.

hectcastro commented on scripts/test:
```diff
@@ -0,0 +1,25 @@
+        usage
+    else
+        docker \
+            run --rm --entrypoint gdalinfo "quay.io/azavea/openjdk-gdal:$GDAL_VERSION-$VARIANT" --version
```
I think that this would be a valuable spot to encode something like:

```
gdalinfo --version
gdalinfo --formats
ogrinfo --formats
```

Commit:
Parameterize OpenJDK version and install proj4 in alpine

* Parameterize the OpenJDK version so that it can be changed and the
  version can be displayed in the image tag
* Install libproj in Alpine images

hectcastro reviewed (approved):
I think that this is in a good spot. Naming stuff doesn't need to be sorted out before approval, and I was able to get the `geotrellis-gdal` tests passing after tweaking the GDAL version reference, like you mentioned. 👍

hectcastro committed:
Use APK testing repository instead of main
